### ISSUE TITLE
AD0001 RemoveUnnecessaryCast crash

### ISSUE THREAD

**Issue by jods4** [dotnet/roslyn#17029]
VS2017 RC, Feb 7th update

One of my source files triggers this:

```
'Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.	***		1	Active	Analyzer 'Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer' threw the following exception:
'Exception occurred with following context:
Compilation: ***
SyntaxTree: G:\proj\***\MassPayments.cs
SyntaxNode: (int)TransferTypeKey.TransferToBeneficiary ... [CastExpressionSyntax]@[2079..2121) (60,25)-(60,67)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Shared.Extensions.ITypeSymbolExtensions.IsEnumType(ITypeSymbol type)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol, Boolean performEquivalenceCheck, Boolean requireNonNullSymbols)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.SymbolInfosAreCompatible(SymbolInfo originalSymbolInfo, SymbolInfo newSymbolInfo, Boolean performEquivalenceCheck, Boolean requireNonNullSymbols)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.SymbolInfosAreCompatible(SymbolInfo originalSymbolInfo, SymbolInfo newSymbolInfo, Boolean requireNonNullSymbols)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.SymbolsAreCompatible(TSyntaxNode originalNode, TSyntaxNode newNode, Boolean requireNonNullSymbols)
   at Microsoft.CodeAnalysis.CSharp.Utilities.SpeculationAnalyzer.ReplacementBreaksBinaryExpression(BinaryExpressionSyntax binaryExpression, BinaryExpressionSyntax newBinaryExpression)
   at Microsoft.CodeAnalysis.CSharp.Utilities.SpeculationAnalyzer.ReplacementChangesSemanticsForNodeLanguageSpecific(SyntaxNode currentOriginalNode, SyntaxNode currentReplacedNode, SyntaxNode previousOriginalNode, SyntaxNode previousReplacedNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemanticsForNode(TSyntaxNode currentOriginalNode, TSyntaxNode currentReplacedNode, TSyntaxNode previousOriginalNode, TSyntaxNode previousReplacedNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemantics(TSyntaxNode currentOriginalNode, TSyntaxNode currentReplacedNode, TSyntaxNode originalRoot, Boolean skipVerificationForCurrentNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemantics()
   at Microsoft.CodeAnalysis.CSharp.Extensions.CastExpressionSyntaxExtensions.IsUnnecessaryCast(CastExpressionSyntax cast, SemanticModel semanticModel, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer.IsUnnecessaryCast(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.RemoveUnnecessaryCast.RemoveUnnecessaryCastDiagnosticAnalyzerBase`1.TryRemoveCastExpression(SemanticModel model, SyntaxNode node, Diagnostic& diagnostic, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.RemoveUnnecessaryCast.RemoveUnnecessaryCastDiagnosticAnalyzerBase`1.<Initialize>b__6_0(SyntaxNodeAnalysisContext nodeContext)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass42_1`1.<ExecuteSyntaxNodeAction>b__1()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze, Nullable`1 info)
-----
'.
```

Here's the offending block of code:
```cs
// Some types worth noting:
// t: dynamic
// TransferTypeKey: enum
foreach (var t in transfers)
  try
  {
    if (t.TYP != (int)TransferTypeKey.TransferToBeneficiary)
      throw new InvalidOperationException($"Transfer is not a transfer to beneficiary, type = {t.TYP}");
  }
```

**Comment by Pilchie** [comment]
Thanks - this looks like it might be the same thing as #9534, but with enough info to maybe fix.

@CyrusNajmabadi or @DustinCampbell care to take a look?

**Comment by DustinCampbell** [comment]
Might be. I took a look at #9534 once but there wasn't enough information there to be actionable. What milestone should this be looked at in? If it's #9534, it's been there a long time.

**Comment by Pilchie** [comment]
Fixed by #17069
